# Set-up

In [2]:
# https://github.com/jayadevbhaskaran/gendered-sentiment

from google.colab import drive
drive.mount('/content/gdrive')  # Upload github files to your drive — this method is persistent.
project_path = '/content/gdrive/My Drive/Colab Notebooks/nlpproject' # Replace 'YourProjectFolder' with your actual project folder name


"""
# Create new directory, downloads the GloVe 6B dataset, rename uppercase like "B" to'b', and unzip all files.
!mkdir -p "{project_path}/glove.6b"  # Create a new directory named 'glove.6b' within your project folder to store GloVe embeddings
!wget http://nlp.stanford.edu/data/glove.6B.zip -O "{project_path}/glove.6b.zip"  # Downloads the GloVe 6B dataset and saves it as 'glove.6b.zip'
!unzip "{project_path}/glove.6b.zip" -d "{project_path}/glove.6b"  # Unzip the downloaded GloVe embeddings zip file into the 'glove.6b' directory
!rename 's/6B/6b/' data/glove.6b/glove.6B.*

!ls "{project_path}/glove.6b"  # List the contents of the 'glove.6b' directory to verify the extractio
"""

Mounted at /content/gdrive


'\n# Create new directory, downloads the GloVe 6B dataset, rename uppercase like "B" to\'b\', and unzip all files.\n!mkdir -p "{project_path}/glove.6b"  # Create a new directory named \'glove.6b\' within your project folder to store GloVe embeddings\n!wget http://nlp.stanford.edu/data/glove.6B.zip -O "{project_path}/glove.6b.zip"  # Downloads the GloVe 6B dataset and saves it as \'glove.6b.zip\'\n!unzip "{project_path}/glove.6b.zip" -d "{project_path}/glove.6b"  # Unzip the downloaded GloVe embeddings zip file into the \'glove.6b\' directory\n!rename \'s/6B/6b/\' data/glove.6b/glove.6B.*\n\n!ls "{project_path}/glove.6b"  # List the contents of the \'glove.6b\' directory to verify the extractio\n'

In [ ]:
import wandb
wandb.login()
wandb.init(project="nlpcapstone", entity="mariamabdullahedu-student")

wandb: Using wandb-core as the SDK backend.  Please refer to https://wandb.me/wandb-core for more information.


<IPython.core.display.Javascript object>

wandb: Logging into wandb.ai. (Learn how to deploy a W&B server locally: https://wandb.me/wandb-server)
wandb: You can find your API key in your browser here: https://wandb.ai/authorize
wandb: Paste an API key from your profile and hit enter, or press ctrl+c to quit:

 ··········


wandb: Appending key for api.wandb.ai to your netrc file: /root/.netrc
wandb: Currently logged in as: mariamabdullahedu (mariamabdullahedu-student). Use `wandb login --relogin` to force relogin


# Actual Code

In [24]:
%cd /content/gdrive/My\ Drive/Colab\ Notebooks/nlpproject

/content/gdrive/My Drive/Colab Notebooks/nlpproject


In [ ]:
!python3 src/baseline.py

0.8268348623853211


In [ ]:
!python3 src/bilstm.py

2024-11-25 15:42:48.499487: I tensorflow/core/util/port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2024-11-25 15:42:48.516971: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:485] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-11-25 15:42:48.537682: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:8454] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-11-25 15:42:48.543996: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1452] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2024-11-25 15:42:48.558822: I tensorflow/core/platform/cpu_feature_guar

In [ ]:
!python3 src/albert_finetune.py

2024-11-23 21:32:13.151863: I tensorflow/core/util/port.cc:153] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2024-11-23 21:32:13.168060: E external/local_xla/xla/stream_executor/cuda/cuda_fft.cc:485] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2024-11-23 21:32:13.189902: E external/local_xla/xla/stream_executor/cuda/cuda_dnn.cc:8454] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2024-11-23 21:32:13.196367: E external/local_xla/xla/stream_executor/cuda/cuda_blas.cc:1452] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2024-11-23 21:32:13.212146: I tensorflow/core/platform/cpu_feature_guar

# Analysis

In [ ]:
!python3 src/analysis.py

Results: [[-5.906212425825358, 7.501185188242566e-09, 0.0345668387793015], [-3.1718725905444343, 0.0016317204812352577, 0.041926803486043274], [1.1685933255210845, 0.24326524140572028, -0.041486174]]


Male:  1182 0.5346869712351946
Female:  601 0.5990016638935108


Profession mean_sentiment female-male
doctor 0.5510329 0.19298714
tailor 0.5010188 0.07261896
baker 0.5839361 0.03975624
secretary 0.55057305 -0.19687605
professor 0.533852 -0.033409
scientist 0.56870306 0.13333184
writer 0.55543166 0.090203464
teacher 0.4526431 -0.31191087
truck driver 0.46130982 -0.14167768
pilot 0.3316312 -0.17408599
lawyer 0.5067767 0.09003383
flight attendant 0.59213984 0.0018737316
nurse 0.4182628 -0.13762197
chef 0.5008215 0.10030687
soldier 0.4912507 -0.31595385
dancer 0.52163684 -0.53959006
gym trainer 0.5522043 0.09706324
mechanic 0.42195368 0.05451402
clerk 0.45122474 0.19907776
bartender 0.62482184 -0.050364792
CONTROL 0.9964181 0.0


noun female-male
He -0.060819007
This boy 0.004403785
This ma

### Sentiment Scores

| **Metric**              | **BERT/Kina Run** | **ALBERT/Mariam Run** | **Diffs/Comment**         |
|--------------------------|-------------------|------------------------|---------------------------|
| Male Sentiment           | 0.535            | 0.535                 | Identical.               |
| Female Sentiment         | 0.599            | 0.599                 | Identical.               |
| Female-Male Difference   | +0.064           | +0.064                | Identical.               |

---

### Profession Analysis

| **Profession**           | **BERT/Kina Run**         | **ALBERT/Mariam Run**      | **Diffs/Comment**                   |
|---------------------------|---------------------------|-----------------------------|-------------------------------------|
| Doctor                   | 0.524 (+0.050)           | 0.551 (+0.193)             | ALBERT shows higher female bias.   |
| Tailor                   | 0.526 (+0.153)           | 0.501 (+0.073)             | BER shows higher female bias.      |
| Secretary                | 0.587 (-0.123)           | 0.551 (-0.197)             | ALBERT shows greater male bias.    |
| Dancer                   | 0.476 (-0.450)           | 0.522 (-0.540)             | ALBERT shows more male bias.       |
| Chef                     | 0.478 (+0.252)           | 0.501 (+0.100)             | BER shows higher female bias.      |
| Nurse                    | 0.429 (-0.058)           | 0.418 (-0.138)             | ALBERT shows more male bias.       |

---

### Control Sentence Evaluation

| **Metric**               | **BERT/Kina Run** | **ALBERT/Mariam Run** | **Diffs/Comment**                   |
|---------------------------|-------------------|------------------------|-------------------------------------|
| Control Mean Sentiment    | 0.999            | 0.996                 | Slightly lower mean for ALBERT.    |
| Control Female-Male Diff. | 0.0              | 0.0                   | Identical.                         |

---
### Noun Analysis

| **Noun**                | **BERT/Kina Run**         | **ALBERT/Mariam Run**       | **Diffs/Comment**                |
|--------------------------|---------------------------|-----------------------------|----------------------------------|
| He                      | +0.0037                  | -0.0608                    | ALBERT shows more male bias.    |
| This man                | -0.1470                  | -0.0129                    | ALBERT shows less male bias.    |
| My father               | -0.1817                  | -0.1024                    | ALBERT reduces male bias.       |
| This widower            | +0.396                   | +0.247                     | BER shows higher female bias.   |

---

### Spinster vs. Bachelor

| **Comparison**             | **BERT/Kina Run**            | **ALBERT/Mariam Run**        | **Diffs/Comment**                  |
|-----------------------------|-----------------------------|-----------------------------|------------------------------------|
| Spinster vs. Bachelor       | -0.101, p = 0.538          | -0.174, p = 0.261          | ALBERT shows stronger male bias, though insignificant. |

###Takeaways:

1. **Male vs. Female Sentiment**: Scores are identical for both models, with females having slightly higher mean sentiment (+0.064).

2. **Profession Analysis**:
   - ALBERT demonstrates higher female bias for professions like "doctor" and "scientist."
   - BERT shows stronger female bias in roles like "chef" and "tailor."
   - ALBERT displays greater male bias for roles such as "dancer" and "secretary."

3. **Control Sentences**:
   - Both models have similar results for control sentences, with identical female-male differences of 0.0. ALBERT has a slightly lower mean sentiment.

4. **Noun Analysis**:
   - ALBERT reduces male bias for nouns like "This man" and "My father" compared to BERT.
   - BERT exhibits higher female bias for nouns like "This widower."

5. **Spinster vs. Bachelor**:
   - ALBERT shows a stronger male bias for "bachelor" compared to "spinster" (difference = -0.174) than BERT (difference = -0.101), though both results remain statistically insignificant.

Overall, ALBERT tends to balance male-female biases in some cases but amplifies them in others, highlighting nuanced differences between the two models.